open-cluster-management/backlog#1820 Perf scal (#554)
open-cluster-management/rhacm-docs: install/perform_scale.md

# Performance and scalability

Red Hat Advanced Cluster Management for Kubernetes is tested to determine some scalability and performance data. The major areas that are tested are cluster scalability and search performance. 

You can use this information to help you plan your environment. 

**Important:** This data is based on the results from a lab environment at the time of testing. Your results might vary, depending on your environment, network speed, and changes to the product. 

## Maximum number of managed clusters

The Advanced Cluster Management for Kubernetes hub cluster provided good performance when managing up to 250 managed clusters. The following table shows the configuration information for the clusters on the Amazon Web Services cloud platfrom that were used to determine the cluster maximums:

| Node | Flavor | vCPU | RAM (GiB) | Disk type | Disk size(GiB)/IOS | Count | Region |
| Master | m5.2xlarge | 8 | 32 | gp2 | 100 | 3 | us-east-1 |
| Worker | m5.2xlarge | 8 | 32 | gp2 | 100 | 3/5 | us-east-1 |

## Search scalability

> The scalability of the Search component depends on the performance of the data store, RedisGraph. The following variables are important to the search performance:

Reviewer: Let's leave "RedisGraph" out. I don't want it to sound like we are blaming a 3rd party.
```suggestion
The scalability of the Search component depends on the performance of the data store. The following variables are important when analyzing the search performance:
```

* Physical memory
* Write throughput (Cache recovery time)
* Query execution time

### Physical memory

> RedisGraph is an in-memory graph database. The memory required is proportional to the number of resources (ConfigMaps, Deployments, Pods, etc) and their relationships in the cluster.

Reviewer: Also removing RedisGraph here.
```suggestion
Search keeps the data in-memory to achieve fast response times. The memory required is proportional to the number of resources (ConfigMaps, Deployments, Pods, etc) and their relationships in the cluster.
```

| Clusters | Kubernetes resources  | Relationships  | Observed size (with simulated data) |
| ------- | --------| -----|----- |
| 1 medium   | 5000 | 9500  | 50 MB |
| 5 medium     | 25,000 | 75,000  | 120 MB | 
| 15 medium     | 75,000 | 20,0000  | 263 MB |
| 30 medium     | 150,000 | 450,000  | 492 MB |
| 50 medium     | 250,000 | 750,000  | 878 MB | 

Table: Physical memory data

> 

Reviewer: ```suggestion
By default, we deploy the datastore with a memory limit of 1Gi.  If you are managing larger clusters, you may need to increase this limit by editing the deployment named `search-prod-xxxxx-redisgraph` in the open-cluster-management namespace.
```

### Write throughput (cache recovery time)

> Most clusters in steady state generate a small number of resource updates. The highest rate of updates happen when the data in RedisGraph is cleared, which causes the remote collectors to get close their full state around the same time.

Reviewer: ```suggestion
Most clusters in steady state generate a small number of resource updates. The highest rate of updates happen when the data in RedisGraph is cleared, which causes the remote collectors to synchronize their full state around the same time.
```

| Clusters | Kubernetes resources  | Relationships  | Average recovery time from simulation |
| ------- | --------| -----|----- |
| 1 medium   | 5000 | 9500  | less than 2 seconds |
| 5 medium     | 25,000 | 75,000  | less than 15 seconds | 
| 15 medium     | 75,000 | 200,000  | 2 minutes and 40 seconds |
| 30 medium     | 150,000 | 450,000  | 5-8 minutes |

Table: Cache recovery time data

**Remember:** Times might increase for clusters that have a slow network connection to the hub.

### Query execution considerations

There are some things that can affect the time that it takes to run and return results from a query. Consider the following items when planning and configuring your environment:

* Searching for a keyword is not efficient.

> * The first search takes longer than later searches because it takes additional time to build the RBAC filter.

Reviewer: ```suggestion
* The first search takes longer than later searches because it takes additional time to gather the user's access rules.
```

> * The length of time to complete a request is proportional to the number of namespaces on the hub cluster.

Reviewer: ```suggestion
* The length of time to complete a request is proportional to the number of namespaces and resources the user is authorized to access.
```

# Performance and scalability

Red Hat Advanced Cluster Management for Kubernetes is tested to determine some scalability and performance data. The major areas that are tested are cluster scalability and search performance. 

You can use this information to help you plan your environment. 

> **Important:** This data is based on the results from a lab environment at the time of testing. Your results might vary, depending on your environment, network throughput, and changes to the product. 

Reviewer 2: Your results might vary, depending on your environment, network throughout, and changes to the product.

Author: Throughput was actually the intent here.  It is basically the speed that your network moves information.  I will tweak it, in case it is confusing.  Thanks!

## Maximum number of managed clusters

The Advanced Cluster Management for Kubernetes hub cluster provided good performance when managing up to 250 managed clusters. 

> | --- | Infrastructure  | OS  | Instance type |  Zone  |  CPU (#vCPU)  |  Memory (GB)  |  Red Hat OpenShift Container Platform version  |  OpenShift Container Platform specification |

Reviewer 3: Not sure we want to put the test system spec in the TP release. When we collect more data for GA, you can probably doc the spec. ie, here're the OCP perf/scale env specs: https://docs.openshift.com/container-platform/4.3/scalability_and_performance/planning-your-environment-according-to-object-maximums.html#cluster-maximums-environment_object-limits. If we want to include the system spec for AWS, something similar to OCP doc, seems nice (Also do not include the test host info. That's to run the performance test tool simulator): 

AWS cloud platform: 
| Node | Flavor | vCUP | RAM (GiB) | Disk type | Disk size(GiB)/IOS | Count | Region |
| Master | m5.2xlarge | 8 | 32 | gp2 | 100 | 3 | us-east-1 |
| Infra | m5.2xlarge | 8 | 32 | gp2 | 100 | 1 | us-east-1 |
| Worker | m5.2xlarge | 8 | 32 | gp2 | 100 | 3/5 | us-east-1 |

Author: Thanks @evelinec ! I made some updates, based on your comments. See if the updates are more the idea that you had in mind.  Thanks!

Reviewer 3: Looks good, thanks @cadawson
